--- a/library/std/src/sys/pal/wasi/os.rs
+++ b/library/std/src/sys/pal/wasi/os.rs
@@ -8,6 +8,9 @@
 use crate::sys::unsupported;
 use crate::{fmt, io};

+#[path = "codepod.rs"]
+mod codepod;
+
 // Add a few symbols not in upstream `libc` just yet.
 pub mod libc {
     pub use libc::*;
@@ -53,40 +56,40 @@
     }
 }

-pub struct SplitPaths<'a>(!, PhantomData<&'a ()>);
+pub struct SplitPaths<'a>(crate::vec::IntoIter<PathBuf>, PhantomData<&'a ()>);

-pub fn split_paths(_unparsed: &OsStr) -> SplitPaths<'_> {
-    panic!("unsupported")
+pub fn split_paths(unparsed: &OsStr) -> SplitPaths<'_> {
+    SplitPaths(codepod::split_paths(unparsed).into_iter(), PhantomData)
 }

 impl<'a> Iterator for SplitPaths<'a> {
     type Item = PathBuf;
     fn next(&mut self) -> Option<PathBuf> {
-        self.0
+        self.0.next()
     }
 }

 #[derive(Debug)]
 pub struct JoinPathsError;

-pub fn join_paths<I, T>(_paths: I) -> Result<OsString, JoinPathsError>
+pub fn join_paths<I, T>(paths: I) -> Result<OsString, JoinPathsError>
 where
     I: Iterator<Item = T>,
     T: AsRef<OsStr>,
 {
-    Err(JoinPathsError)
+    codepod::join_paths(paths).ok_or(JoinPathsError)
 }

 impl fmt::Display for JoinPathsError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        "not supported on wasm yet".fmt(f)
+        "path segment contains separator `:`".fmt(f)
     }
 }

 impl crate::error::Error for JoinPathsError {}

 pub fn current_exe() -> io::Result<PathBuf> {
-    unsupported()
+    codepod::current_exe()
 }

 #[allow(dead_code)]
@@ -95,11 +98,11 @@
 }

 pub fn temp_dir() -> PathBuf {
-    panic!("no filesystem on wasm")
+    codepod::temp_dir()
 }

 pub fn home_dir() -> Option<PathBuf> {
-    None
+    codepod::home_dir()
 }

 pub fn exit(code: i32) -> ! {
@@ -107,7 +110,7 @@
 }

 pub fn getpid() -> u32 {
-    panic!("unsupported");
+    codepod::getpid()
 }

 #[doc(hidden)]
